users/user-guide-query/multi-stage-query/operator-types/mailbox-receive.md

```diff
@@ -0,0 +1,110 @@
+---
+description: >-
+  Describes the mailbox receive operator in the multi-stage query engine.
+---
+
+# Mailbox receive operator
+
+The mailbox receive operator is the operator that receives the data from the mailbox send operator.
+This is not an actual relational operator but a Pinot extension used to receive data from other stages.
+
+## Implementation details
+
+Stages in the multi-stage query engine are executed in parallel by different workers.
+Workers send data to each other using mailboxes.
+The number of mailboxes depends on the send operator parallelism, the receive operator parallelism and the distribution
+being used.
+At worse, there is one mailbox per worker pair, so if the upstream send operator has a parallelism of S and the receive 
+operator has a parallelism of R, there will be S * R mailboxes.
+
+By default, these mailboxes are GRPC channels, but when both workers are in the same server, they can use shared memory
+and therefore a more efficient on heap mailbox is used.
+
+The mailbox receive operator pulls data from these mailboxes and sends it to the downstream operator.
+
+### Blocking nature
+
+The mailbox receive operator is a streaming operator.
+It emits the blocks of rows as soon as they are received from the upstream operator.
+
+It is important to notice that the mailbox receive operator tries to be fair when reading from multiple workers.
```

Comment:
what does this mean that it tries to be fair? what exactly does it do?

Reply:
If two workers send data to the same mailbox receive, then the implementation does not simply read all data from one of the workers and then read the other. Instead we read some blocks from the first one and then read from the other (even if we have more data to read from the first one). This is a decision taken long ago to try to do not block workers when several of them are emitting values.